rust/arrow/src/array/transform/variable_size.rs

```diff
@@ -43,33 +47,33 @@ pub(super) fn build_extend<T: OffsetSizeTrait>(array: &ArrayData) -> Extend {
         // fast case where we can copy regions without null issues
         Box::new(
             move |mutable: &mut _MutableArrayData, _, start: usize, len: usize| {
-                let mutable_offsets = mutable.buffer::<T>(0);
-                let last_offset = mutable_offsets[mutable_offsets.len() - 1];
-                // offsets
-                let buffer = &mut mutable.buffers[0];
+                let offset_buffer = &mut mutable.buffer1;
+                let values_buffer = &mut mutable.buffer2;
+
+                // this is safe due to how offset is built. See details on `get_last_offset`
+                let last_offset = unsafe { get_last_offset(offset_buffer) };
```

Comment:
I find  use of unsafe in order to call `offset_buffer` somewhat confusing. I suspect you are trying to follow the unsafe guidelines and ensure it is clear where unsafe is being used. 

However, in this case the only thing the caller can do is trust that  the `MutableBuffer` it was passed in was created correctly. Forcing callers to say `unsafe`  in order for the call to `get_last_offset` even though they can do nothing to ensure/validate things are safe or not seems unnecessarily confusing to me

I would personally suggest making `get_last_offset`  an associated function, such as `MutableBuffer::get_last_offset`

And then change calls such as this to

```
let last_offset = offset_buffer.get_last_offset();
```

Reply:
I get your point.

`get_last_offset` must only be used in offset buffers, i.e. buffers whose bytes represent a i32 or i64 and were specifically built from those types. Doing so in other buffers is undefined behavior (even with the safeguards of using `align_to`). We can remove the `unsafe` mark from `get_last_offset`, though.

My proposal is that we refactor the `src/transform` code so that it has a struct specific for each array type (that implements some trait for `dyn` support). This will allow `last_offset` to be stored in the array-specific struct, thereby avoiding this problem altogether (of having to read bytes written to the `MutableBuffer`).

Reply:
Sounds like a good plan to me